make-base/utils.mk

```diff
@@ -0,0 +1,99 @@
+.PHONY: always
+
+# Function that find project directories by a file that must exist inside of it
+define find_projects_containing
+$(sort \
+	$(call map,readlink, \
+		$(filter-out $(shell dirname $(firstword $(wildcard $(addsuffix /monoetna,.. ../.. ../../..))))/%, \
+			$(wildcard \
+				$(addsuffix /*/$(1),. .. ../.. ../../..) \
+				$(addsuffix /docker/*/$(1),. .. ../.. ../../..) \
+				$(addsuffix /swarm/*/$(1),. .. ../.. ../../..) \
+				$(addsuffix /etna/packages/*/$(1),. .. ../.. ../../..) \
+			) \
+		) \
+	) \
+)
+endef
+
+define find_project_containing
+$(call map,dirname,$(filter %/$(1)/$(2),$(call find_projects_containing,$(2))))
+endef
+
+define find_project_file
+$(addsuffix /$(2),$(call find_project_containing,$(1),$(2)))
+endef
+
+map = $(foreach a,$(2),$(call $(1),$(a)))
+readlink = $(shell readlink -m $(1))
+dirname = $(shell dirname $(1))
+
+define find_updated_sources
+$(shell $(call find_project_file,build_support,find-updated-sources) $(1))
+endef
+define updated_build_files
+$(call find_updated_sources,$(call find_project_containing,build_support,build_image))
+endef
+
+images=$(shell cat $(1) | grep FROM | xargs -n2 echo | cut -d ' ' -f2)
+compose_images=$(shell cat $(1) | grep image: | xargs -n2 echo | cut -d ' ' -f2)
+
+# ( dockerfile )
+define buildable_dependent_image_dockerfiles
+$(foreach image,$(call images,$(1)),$(call find_project_file,$(image),Dockerfile))
+endef
+
+# ( composefile )
+define buildable_dependent_compose_image_dockerfiles
+$(foreach image,$(call compose_images,$(1)),$(call find_project_file,$(image),Dockerfile))
+endef
+
+seen_images:=
+# (dockerfile folder, dependent image dockerfiles)
+define image_target2
+ifeq ($(filter $(1)!,$(seen_images)),)
+$(eval seen_images:=$(seen_images) $(1)!)
+$(shell if ! test -e $(1)/image.marker; then touch -t 0001011000 $(1)/image.marker; fi)
```

Comment:
Because make's logic to build something is to first check if the file exist, and run its target if it does not, we want to ensure that there is always a marker file (with the oldest age possible), so that if no source files are found to be updated recently, the marker already exists and no work is done.